src/span.h

```diff
@@ -61,12 +67,6 @@ class Span
     template <typename O, typename std::enable_if<std::is_convertible<O (*)[], C (*)[]>::value, int>::type = 0>
     constexpr Span(const Span<O>& other) noexcept : m_data(other.m_data), m_size(other.m_size) {}
 
-    /** Default copy constructor. */
-    constexpr Span(const Span&) noexcept = default;
-
-    /** Default assignment operator. */
-    Span& operator=(const Span& other) noexcept = default;
```

Comment:
in commit 397e5eda38144147330163e7a51413ac01d1d6f5

It looks like std::span has no move constructor (https://en.cppreference.com/w/cpp/container/span/span), so shouldn't we do the same?

Reply:
Hmm. I'm seeing conflicting info on this.

According to the (official?) revision 7 draft, the move constructor was removed in r6. If we were to follow that, I believe we would drop our fun rvalue ctor.

I'm not sure what draft version cppreference is referencing, but I believe it must be an old one.

libc++ for example has implemented r7: https://reviews.llvm.org/D49338.

Will explore more.

Reply:
looking at the source code, it shows they are using `= default`, no?

https://github.com/llvm/llvm-project/blame/6fafde0387229d6656faee41100b73615343819a/libcxx/include/span#L70

Reply:
@MarcoFalke Yes, but they also don't have a universal reference constructor like we do.

The problem is that return values are passed through a move constructor if they exist, before copy constructors. See [here](https://en.cppreference.com/w/cpp/language/return) under "Automatic move from local variables and parameters".

Because the default move constructor is removed by adding the default constructor/assignment operator, our return values are passed through our universal constructor as rvalues. That's really not a problem except that we want to minimize what's going through that constructor so that we can apply the lifetimebound attribute to it.

( Bonus TIL: apparently c++17 does away with this return value path altogether when possible: "_If expression is a prvalue, the result object is initialized directly by that expression. This does not involve a copy or move constructor when the types match_" )

So there are 2 possible solutions for us (if you're willing to accept that there's something to be solved)

1. remove the default ctor/assignment operator as I've done here, thus enabling a trivial move constructor for return values to pass through
2. remove our universal reference constructor and replace it with something less greedy

Your question made me look up the current c++20 draft spec, where I noticed that the universal reference constructor has been removed in a newish revision. So what's on cppreference is actually stale.

But that gives us an easy answer to the question above, we should just update our implementation to match the latest draft spec. I'm going to close this and open a new PR with that approach, since it's sufficiently different from what's here.